server/src/main/java/org/candlepin/model/Product.java

```diff
@@ -397,58 +560,143 @@ public boolean hasContent(String contentId) {
+
+                        return false;
+                    }
+                }
+
+                if (!Util.collectionsAreEqual(this.productContent, that.productContent)) {
+                    if (!(this.productContent == null && that.productContent.size() == 0) &&
+                        !(that.productContent == null && this.productContent.size() == 0)) {
 
-        // TODO: Maybe checking the UUID would be better here...?
-        return (this.getOwner() != null ? this.getOwner().equals(another.getOwner()) :
-            another.getOwner() == null) && getId().equals(another.getId()) &&
-            name.equals(another.getName());
-    }
+                        return false;
+                    }
+                }
 
-    @Override
-    public Object clone() {
-        return new Product(this);
+                if (!Util.collectionsAreEqual(this.dependentProductIds, that.dependentProductIds)) {
+                    if (!(this.dependentProductIds == null && that.dependentProductIds.size() == 0) &&
+                        !(that.dependentProductIds == null && this.dependentProductIds.size() == 0)) {
+
+                        return false;
+                    }
+                }
+            }
+
+            return equals;
+        }
+
+        return false;
     }
 
     @Override
     public int hashCode() {
-        if (id != null) {
-            return id.hashCode() * 31;
+        // This must always be a subset of equals
+        HashCodeBuilder builder = new HashCodeBuilder(37, 7)
+            .append(this.id)
+            .append(this.name)
+            .append(this.multiplier);
+
+        // Impl note:
+        // Because we handle the collections specially in .equals, we have to do the same special
+        // treatment here to ensure our output doesn't give us wonky results when compared to the
+        // output of .equals
+
+        if (this.attributes != null && this.attributes.size() > 0) {
+            builder.append(this.attributes);
+        }
+
+        try {
+            if (this.productContent != null && this.productContent.size() > 0) {
+                builder.append(this.productContent);
+            }
+
+            if (this.dependentProductIds != null && this.dependentProductIds.size() > 0) {
+                builder.append(this.dependentProductIds);
+            }
         }
-        return 31;
+        catch (LazyInitializationException e) {
+            // One of the above collections (likely the first) has not been initialized and we're
+            // not able to fetch them. We still need a hashCode, and the caller is likely to run
+            // into this exception in the very near future, but we still need to generate
+            // something here. We'll treat it as if they were empty (and not added).
+
+            // Note that this only applies to lazy collections, so the product attributes are safe
+            // to check. ...for now.
+        }
```

Comment:
@crog REMINDER: If the collections are no longer lazy, then we should remove this guard since we are relying on hascode for versioning.